src/nvim/channel.c

```diff
@@ -0,0 +1,499 @@
+    }
+    rettv->vval.v_number = proc->status;
+    channel_decref(data);
+    return false;
+  }
+  xfree(cmd);
+
+
+  if (data->is_rpc) {
+    // the rpc takes over the in and out streams
+    rpc_start(data);
+  } else {
+    wstream_init(&proc->in, 0);
+    if (has_out) {
+      callback_reader_start(&data->on_stdout);
+      rstream_init(&proc->out, 0);
+      rstream_start(&proc->out, on_job_stdout, data);
+    }
+  }
+
+  if (has_err) {
+    callback_reader_start(&data->on_stderr);
+    rstream_init(&proc->err, 0);
+    rstream_start(&proc->err, on_job_stderr, data);
+  }
+  rettv->vval.v_number = (varnumber_T)data->id;
+  return true;
+}
+
+
+uint64_t channel_connect(bool tcp, const char *address,
+                         bool rpc, CallbackReader on_output,
+                         int timeout, const char **error)
+{
+  // TODO: test mixed mode (server RPC, client bytes)
+  if (!tcp && rpc) {
+    char *path = fix_fname(address);
+    if (rpc && server_owns_pipe_address(path)) {
+      // avoid deadlock
+      xfree(path);
+      return channel_create_internal();
+    }
+    xfree(path);
+  }
+
+  Channel *channel = channel_alloc(kChannelTypeSocket);
+  if (!socket_connect(&main_loop, &channel->stream.socket,
+                      tcp, address, timeout, error)) {
+    // TODO: early detroy that doesn't consume a channel id?
+    channel_decref(channel);
+    return 0;
+  }
+
+  channel_incref(channel);  // close channel only after the stream is closed
+  channel->stream.socket.internal_close_cb = close_cb;
+  channel->stream.socket.internal_data = channel;
+  if (rpc) {
+    rpc_start(channel);
+  } else {
+    channel->on_stdout = on_output;
+    callback_reader_start(&channel->on_stdout);
+    wstream_init(&channel->stream.socket, 0);
+    rstream_init(&channel->stream.socket, 0);
+    rstream_start(&channel->stream.socket, on_socket_output, channel);
+  }
+
+  return channel->id;
+}
+
+/// Creates an API channel from stdin/stdout. This is used when embedding
+/// Neovim
+uint64_t channel_from_stdio(bool rpc, CallbackReader on_output, const char **error)
+{
+  if (!headless_mode) {
+    *error = _("Stdio channel can only be opened in headless mode");
+    return 0;
+  }
+
+  if (did_stdio) {
+    *error = _("Stdio channel was already open");
+    return 0;
+  }
+  Channel *channel = channel_alloc(kChannelTypeStdio);
+  channel_incref(channel);  // stdio channels are only closed on exit
+  // read stream
+  rstream_init_fd(&main_loop, &channel->stream.stdio.in, 0, CHANNEL_BUFFER_SIZE);
+  wstream_init_fd(&main_loop, &channel->stream.stdio.out, 1, 0);
+
+  if (rpc) {
+    rpc_start(channel);
+  } else {
+    channel->on_stdout = on_output;
+    callback_reader_start(&channel->on_stdout);
+    rstream_start(&channel->stream.socket, on_stdio_input, channel);
+  }
+
+  return channel->id;
+}
+
+/// @param data will be consumed
+size_t channel_send(uint64_t id, char* data, size_t len, const char **error)
```

Comment:
`fwrite()` only returns the number of items written (here `1`). That number is also not equal the number of bytes written if `len != 1`.